### PR TITLE
[Merged by Bors] - feat: deduplicate lemmas used by rw?

### PR DESCRIPTION
**Pull request by kim-em**
Solving request 1. from @b-mehta's [zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Duplicates.20in.20rw.3F/near/381963606) post.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: test/rewrites.lean [hunk 1]
```diff
@@ -130,3 +132,9 @@ info: Try this: rw [@AddCommMonoidWithOne.add_comm]
 #guard_msgs in
 example (n : ℕ) : let y := 3; n + y = 3 + n := by
   rw?!
+
+-- Check that the same lemma isn't used multiple times.
+-- This used to report e.g. two redundant copies of `Nat.add_assoc`.
+lemma test (n m : ℕ) : n + m + 1 = m + n + 1 := by
```
Comment (alexjbest) [inline test/rewrites.lean:138]: Can we add a guard msgs here or something to make this an actual test test? (Maybe abstracting away from nat so not too many results are returned

**Comment by semorrison** [comment]
> LGTM, I guess the performance is still ok?

I didn't explicitly test this. The test file still seems fast, and the method I used should be fine.

File: test/rewrites.lean (hunk 1 above)
Comment (semorrison) [reply test/rewrites.lean]: Done!

**Comment by alexjbest** [comment]
maintainer merge

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**